ducktape/mark/_mark.py

```diff
@@ -0,0 +1,359 @@
+    """
+    def parametrizer(f):
+        Mark.mark(f, Parametrize(**kwargs))
+        return f
+    return parametrizer
+
+
+def ignore(*args, **kwargs):
+    """
+    Test method decorator which signals to the test runner to ignore a given test.
+
+    Example::
+
+        When no parameters are provided to the @ignore decorator, ignore all parametrizations of the test function
+
+        @ignore  # Ignore all parametrizations
+        @parametrize(x=1, y=0)
+        @parametrize(x=2, y=3)
+        def the_test(...):
+            ...
+
+    Example::
+
+        If parameters are supplied to the @ignore decorator, only ignore the parametrization with matching parameter(s)
+
+        @ignore(x=2, y=3)
+        @parametrize(x=1, y=0)  # This test will run as usual
+        @parametrize(x=2, y=3)  # This test will be ignored
+        def the_test(...):
+            ...
+    """
+    if len(args) == 1 and len(kwargs) == 0:
+        # this corresponds to the usage of the decorator with no arguments
+        # @ignore
+        # def test_function:
+        #   ...
+        Mark.mark(args[0], IgnoreAll())
+        return args[0]
+
+    # this corresponds to usage of @ignore with arguments
+    def ignorer(f):
+        Mark.mark(f, Ignore(**kwargs))
+        return f
+
+    return ignorer
+
+
+def _inject(*args, **kwargs):
+    """Inject variables into the arguments of a function or method.
+    This is almost identical to decorating with functools.partial, except we also propagate the wrapped
+    function's __name__.
+    """
+
+    def injector(f):
+        assert callable(f)
+
+        @functools.wraps(f)
+        def wrapper(*w_args, **w_kwargs):
+            return functools.partial(f, *args, **kwargs)(*w_args, **w_kwargs)
+
+        wrapper.args = args
+        wrapper.kwargs = kwargs
+        wrapper.function = f
+
+        return wrapper
+    return injector
+
+
+class MarkedFunctionExpander(object):
+    """This class helps expand decorated/marked functions into a list of test context objects. """
+    def __init__(self, session_context=None, module=None, cls=None, function=None):
+        self.session_context = session_context
+        self.module = module
+        self.cls = cls
+        self.function = function
+
+    def expand(self, test_parameters=None):
+        """Inspect self.function for marks, and expand into a list of test context objects useable by the test runner.
+        """
+        f = self.function
+        if not hasattr(f, "marks"):
+            return [TestContext(self.session_context, self.module, self.cls, self.function)]
+
+        expanded = []
+        parametrize_marks = [m for m in f.marks if m.name == PARAMETRIZED.name]
+        matrix_marks = [m for m in f.marks if m.name == MATRIX.name]
+        ignore_marks = [m for m in f.marks if m.name == IGNORE.name]
+
+        ignore_list = [m.injected_args for m in ignore_marks]
+        ignore_all = None in ignore_list
+
+        if test_parameters is not None:
```

Comment:
This seems like it's going to get very messy over time as we have more annotations. We've basically centralized the entire process of expanding the set of tests into one class/method that has to understand everything that all annotations do.

Is there a reason we can't keep each of these blocks in their respective `Mark` classes and only have one smaller method here that manages the high-level process of generating the initial singular `TestContext` for the unmarked test function, then dispatching a `list` of `TestContext`'s to each `Mark`, which generates an output `list` of `TestContext`'s that it derives from the input list?

It seems like this should compose well, cleanly separate each annotation, and discourages dependencies between annotations that cannot be expressed simply via a `TestContext` -- basically, if different annotations want to interact, they have to agree to a clear contract for how to exchange information about what they are doing via `TestContext` attributes alone (which, of course, are extensible by adding attributes, as is done here with, e.g., `ignore`).

In theory this also provides a cleaner mechanism for extending the system with project-specific annotations, though clearly you also need the right hooks into other parts of the system for that to work (e.g. you couldn't support a user-defined `ignore` annotation without also providing a way to hook into the runner/reporter to actual implement the part that checks the `ignore` flag and takes the right action).